### PR TITLE
Upgrading knative to v1beta1

### LINKED ISSUE #190: Upgrade KFService to use Knative v1beta1 API
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by johnugeorge**
**What this PR does / why we need it**:
This PR upgrade knative to v1beta1 API in release version 0.7.1. There are few changes in Knative v1beta1 api(compared to v1alpha1).  Corresponding changes are incorporated

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #190 
Related: #221 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/238)

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -55,7 +56,8 @@ version="v1.4.7"
 [[override]]
   name = "github.com/knative/pkg"
   # HEAD as of 2019-04-19
```
Comment (ellis-bigelow) [inline Gopkg.toml]: Can you update this comment?

File: Gopkg.toml [hunk 2]
```diff
@@ -45,7 +45,8 @@ version="v1.4.7"
 
 [[constraint]]
   name = "github.com/knative/serving"
-  revision = "a44e3578ce27cc1455b3e1c7f74b09b82dda57f5"
+  #revision = "a44e3578ce27cc1455b3e1c7f74b09b82dda57f5"
```
Comment (ellis-bigelow) [inline Gopkg.toml]: commented out?

File: Gopkg.toml [hunk 3]
```diff
@@ -55,7 +56,8 @@ version="v1.4.7"
 [[override]]
   name = "github.com/knative/pkg"
   # HEAD as of 2019-04-19
-  revision = "3c8c4a93547f62b542013c273068f51875b5c942"
+  #revision = "3c8c4a93547f62b542013c273068f51875b5c942"
```
Comment (ellis-bigelow) [inline Gopkg.toml]: dd?

File: pkg/apis/serving/v1alpha1/kfservice_status.go [hunk 4]
```diff
@@ -88,9 +88,10 @@ func (ss *KFServiceStatus) PropagateCanaryConfigurationStatus(canaryConfiguratio
 }
 
 // PropagateRouteStatus propagates route's status to the service's status.
-func (ss *KFServiceStatus) PropagateRouteStatus(rs *knservingv1alpha1.RouteStatus) {
-	ss.URL = rs.Domain
-
+func (ss *KFServiceStatus) PropagateRouteStatus(rs *knservingv1beta1.RouteStatus) {
+	if rs.URL != nil {
+		ss.URL = rs.URL.Host
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_status.go]: Can we ducktype our status to be like theirs? I expect we also want to have scheme, etc.

File: pkg/apis/serving/v1alpha1/kfservice_validation_test.go [hunk 5]
```diff
@@ -175,7 +175,7 @@ func TestCustomBadFields(t *testing.T) {
 			StdinOnce: true,
 		},
 	}
-	g.Expect(kfsvc.ValidateCreate()).Should(gomega.MatchError("Custom container validation error: must not set the field(s): name, stdin, stdinOnce"))
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha1/kfservice_validation_test.go]: Why did you remove name here?

File: pkg/controller/kfservice/resources/knative/configuration.go [hunk 6]
```diff
@@ -83,30 +83,31 @@ func (c *ConfigurationBuilder) CreateKnativeConfiguration(name string, metadata
...
+					// we may need to expose this field in future
+					TimeoutSeconds: &constants.DefaultTimeout,
+					PodSpec: corev1.PodSpec{
+						ServiceAccountName: modelSpec.ServiceAccountName,
+						Containers: []corev1.Container{
+							*modelSpec.CreateModelServingContainer(metadata.Name, c.frameworksConfig),
 						},
 					},
-					Container: modelSpec.CreateModelServingContainer(metadata.Name, c.frameworksConfig),
+					//Container: modelSpec.CreateModelServingContainer(metadata.Name, c.frameworksConfig),
```
Comment (ellis-bigelow) [inline pkg/controller/kfservice/resources/knative/configuration.go]: Don't forget to clean this up.

File: pkg/apis/serving/v1alpha1/kfservice_status.go (hunk 4 above)
Comment (johnugeorge) [reply pkg/apis/serving/v1alpha1/kfservice_status.go]: So, you mean changing our type to URL too?

File: pkg/apis/serving/v1alpha1/kfservice_validation_test.go (hunk 5 above)
Comment (johnugeorge) [reply pkg/apis/serving/v1alpha1/kfservice_validation_test.go]: we are using Knative `ValidateContainer` which has `name` in the allowed list  https://github.com/knative/serving/blob/release-0.7/pkg/apis/serving/fieldmask.go#L194 (and was disallowed earlier)

**Comment by johnugeorge** [comment]
@ellis-bigelow added all changes. I couldn't find a way to separate vendor code from code changes as they go together.  Please take a look

File: pkg/apis/serving/v1alpha1/kfservice_types.go [hunk 7]
```diff
@@ -99,7 +100,7 @@ type CustomSpec struct {
 // KFServiceStatus defines the observed state of KFService
 type KFServiceStatus struct {
 	duckv1beta1.Status `json:",inline"`
-	URL                string                  `json:"url,omitempty"`
+	URL                *apis.URL               `json:"url,omitempty"`
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha1/kfservice_types.go]: Looks knative URL is just an alias of net.URL(https://github.com/knative/pkg/blob/master/apis/url.go#L28), also we require the types to have `+k8s:openapi-gen=true` annotation now to generate python sdk and it looks that `net.URL` does not have the annotation. @jinchihe Can we double check if this is an issue ?

Comment (jinchihe) [reply pkg/apis/serving/v1alpha1/kfservice_types.go]: @yuzisun @johnugeorge Yes, seems that's a problem. But I think we can go ahead to upgrade knative, I will try to upgrade openapi and sdk once the PR merged. I will log a ticket to trace that, thanks.

**Comment by ellis-bigelow** [comment]
/hold
Make sure you get this: https://github.com/kubeflow/kfserving/blob/master/pkg/apis/serving/v1alpha1/kfservice_types.go#L119

**Comment by johnugeorge** [comment]
@ellis-bigelow it is correctly printed as string


Screen Shot 2019-07-16 at 10 17 06 PM

**Comment by ellis-bigelow** [comment]
> @ellis-bigelow it is correctly printed as string
> 
> Screen Shot 2019-07-16 at 10 17 06 PM

Discussed offline. Cool!

/lgtm
/approve

**Comment by johnugeorge** [comment]
/hold 
Verifying an error seen in my testing

**Comment by johnugeorge** [comment]
` kubectl apply -f  docs/samples/tensorflow/tensorflow.yaml` works without any issue.  Routes are populated and curl predict request succeeds. 

However, when on applying `tensorflow-canary.yaml` again,  i am seeing following schema errors randomly. 

`{"level":"error","ts":1563306657.0418444,"logger":"kfserving-controller","msg":"Failed to reconcile","error":"Internal error occurred: no kind \"Route\" is registered for version \"serving.knative.dev/v1beta1\" in scheme \"k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:51\"","stacktrace":"github.com/kubeflow/kfserving/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/johnugeo/GO-ROOT/src/github.com/kubeflow/kfserving/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/kubeflow/kfserving/pkg/controller/kfservice.(*ReconcileService).Reconcile\n\t/Users/johnugeo/GO-ROOT/src/github.com/kubeflow/kfserving/pkg/controller/kfservice/kfservice_controller.go:156\ngithub.com/kubeflow/kfserving/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/johnugeo/GO-ROOT/src/github.com/kubeflow/kfserving/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215`

**Comment by johnugeorge** [comment]
Quite strange as initial routes are added correctly. Have you seen this error before ?  
@yuzisun @ellis-bigelow

**Comment by ellis-bigelow** [comment]
Did you try fully deleting everything and re-deploying including recreating the resources? I'm guessing this is a configuration staleness issue that exists generally, rather than being directly related to the canary.

It looks like either the v1beta1 CRD isn't applied to the cluster OR the controller doesn't have RBAC to CRUD them.

**Comment by johnugeorge** [comment]
Closing this as #251 is merged. 
Knative v1beta1 api will be enabled when k8s 1.14+ is available for general use.

/close
